### DISCUSSION TITLE
Optimal_laptime outputs

### DISCUSSION THREAD

**Discussion by pierre-trott** [juanmanzanero/fastest-lap#19] [Q&A] [answered]
I am using the example code to run an optimal_laptime simulation.
I see how the outputs are defined through the options and then the values are retrieved through:
`fastest_lap.download_vector("run/output_value")`

Is there a place where I can find all the possible outputs from the simulation? Inside a C++ file perhaps?
I guess this will be available in the documentation in the future.
If they are available, I am particularly interested in:
- info about what's happening at each tire (speeds, accelerations, energies...)
- distance and time channels (I thought s might be time channel but max value doesn't make sense for laptime I'm targeting)
- braking
- lateral and longitudinal accelerations
- x, y position of the car in latitude/longitude to export Ideal line as kml to Google Earth. If not available, where can I find the code that does the conversion lat/long -> x/y?
 
I see you have some of these details in the videos so I'm guessing it is somewhere.

Thanks in advance!

**Comment by juanmanzanero** [comment]
Hi Pierre, I have moved your query to the discussions site. I will answer shortly

**Comment by juanmanzanero** [comment, marked as answer]
Hi,

So, the documentation is coming along but it is still not ready.

In the mean time, you can look at the code: [fastestlapc.cpp](https://github.com/juanmanzanero/fastest-lap/blob/main/src/main/c/fastestlapc.cpp) line 210 and beyond, these are the names of the variables that can be exported. If you have special variable request that are not cited there you can open an issue. Variable names are provisional, they will change soon to something like `chassis.u` instead of `u`. I will notify about it in the next release (feel free to 'watch' the repository to stay up to date with upcoming changes)

- `s` is the arc length (distance traveled along the circuit) not the time. The time is simply `time` and the laptime is `laptime`
- `throttle` is both throttle (>0) and brake (<0)
- lat/long accelerations are exported by the names of `ax` (long) and `ay` (lat).
- the formulas to convert (x,y) to (long,lat) are written in the track XML files, e.g. [catalunya_2022_adapted.xml](https://github.com/juanmanzanero/fastest-lap/blob/main/database/tracks/catalunya_2022/catalunya_2022_adapted.xml)
```xml
    <GPS_parameters>
        <!-- x = earth_radius.cos(reference_latitude).(longitude - origin_longitude) -->
        <!-- y = earth_radius.(latitude - origin_latitude) -->
        <origin_longitude units="deg">2.2630628832470472</origin_longitude>
        <origin_latitude units="deg">41.572323326229437</origin_latitude>
        <earth_radius units="m">6378388</earth_radius>
        <reference_latitude units="deg">41.572323326229437</reference_latitude>
    </GPS_parameters>
```

**Reply by pierre-trott** [reply]
Thank you for this information!

**Reply by pierre-trott** [reply]
Hi,

I'm trying to get lateral and longitudinal acceleration out of the simulation.
I added
`options += "            <ax/>"`
`options += "            <ax/>"`

but when I add those two lines in the options, the optimal_laptime() doesn't finish.
I see all the iterations in the console but after the: `EXIT: Optimal Solution Found.`, I have an error.
I see the `property_name = 'ax'` in [fastestlapc.ccp](https://github.com/juanmanzanero/fastest-lap/blob/main/src/main/c/fastestlapc.cpp) so I expected it to work like the other variables.

Do I have to have something else in the options to get the lateral accelerations?
I have ax and ay inside variables, is that okay?

Thanks

**Reply by juanmanzanero** [reply]
Hello! Could be that you are typing `<ax/>` twice instead of `<ax/> <ay/>`. I don't know if it is a typo in this message or in the script.

**Reply by pierre-trott** [reply]
Sorry this is a typo in the message, not the script.
This is what I have:
![image](https://user-images.githubusercontent.com/33380319/177321781-0fbe0faa-a2a9-4722-8350-34285065522e.png)

This worked fine until I added `<ax/> <ay/>`.

**Reply by juanmanzanero** [reply]
Yes, you are right.  The accelerations are not optimal_laptime outputs but there can be computed later by calling a postprocess function which is not on python, yet. I will open an issue! Apologies

**Reply by pierre-trott** [reply]
Okay, thanks!
